### ISSUE TITLE
Idea: allow passing strategies for remaining arguments to a @given'd test

### ISSUE THREAD

**Issue by Zac-HD** [HypothesisWorks/hypothesis-python#1146]
That is:

```python
@given(a=integers(min_value=1))
def test_addition(a, b):
   assert a + b > 0

test(0)  # passes
test_addition(-2)  # fails; Falsifying Example: test_addition(a=1. b=-2)
# proposed:
test(integers())  # fails; Falsifying Example: test_addition(a=1. b=-1)
```

Turns out it would be really easy to hook additional strategies into the Hypothesis run (modifying `hypothesis.core.process_arguments_to_given`).

This could be really useful for e.g. Pytest fixtures - just return a strategy! - but would change existing behaviour of tests which are passed a strategy as an argument (although easy to work around, that's technically a breaking change).  

We would presumably allow zero-argument given and raise an error if no strategy is supplied, which opens up interesting design questions around using strategies as default values for arguments to test functions.  (*Is* it an error to supply no strategy at all, if one could have been given?  What would we do then?)

TLDR: design idea would make `@given` more flexible - but is that a good thing?

**Comment by dchudz** [comment]
Can you say more about why it's useful for a fixture to return a strategy? (As opposed to defining the strategy and using it in `@given` directly?)

I'm asking largely because I think the way I would *want* `@given` to work with fixtures is to let you decorate fixtures with `@given`. I don't know that we'll ever be able to do it, but I think anything with fixtures & strategies that isn't that makes me a little nervous users will think we're supporting that. (And adds confusion if we ever do somehow support it.)

**Comment by DRMacIver** [comment]
I'm kinda -1 on this. I'd like `@given` to get *less* complicated over time rather than more! 

I'm not set on it as a plan, but ideally in the long-run we'd move `@given` in line with how Hypothesis works/is going to work in other languages, prioritising the inline style over the current approach.

**Comment by Zac-HD** [comment]
> Can you say more about why it's useful for a fixture to return a strategy? (As opposed to defining the strategy and using it in @given directly?)

The effect after the test is called would be identical, so this would "just" add a different style you can use to get the same results.

> I think the way I would want `@given` to work with fixtures is to let you decorate fixtures with `@given`.

In this proposal, you would instead decorate them with `st.composite` - fixtures are straetgies, not tests!

> I'm kinda -1 on this. I'd like @given to get less complicated over time rather than more!  I'm not set on it as a plan, but ideally in the long-run we'd move @given in line with how Hypothesis works/is going to work in other languages, prioritising the inline style over the current approach.

FWIW I expect this to be a 2-4 line diff to implement, including error handling (but not tests or checks for easy mistakes).  It would allow more complex *usage*, but `@given` itself would barely change.

Re: long term - I've always found it easier to debug given arguments than with inline draws, so I'm not (currently) keen on that plan.  Open to further discussion elsewhere, but in the short term I don't think this would materially change our deprecation pathway.
